0.17.0: pytest fails in 4 units: cpp mixed-linkage

I'm packaging your module as an rpm package so I'm using the typical PEP517 based build, install and test cycle used on building packages from non-root account.
- `python3 -sBm build -w --no-isolation`
- because I'm calling `build` with `--no-isolation` I'm using during all processes only locally installed modules
- install .whl file in </install/prefix> using `installer` module
- run pytest with $PYTHONPATH pointing to sitearch and sitelib inside </install/prefix>
- build is performed in env which is *`cut off from access to the public network`* (pytest is executed with `-m "not network"`)

<details>
<summary>Here is pytest output:</summary>

```console
Package                       Version
----------------------------- -----------
alabaster                     0.7.16
Babel                         2.14.0
build                         1.1.1
charset-normalizer            3.3.2
clang                         18.1.2
distro                        1.9.0
docutils                      0.20.1
exceptiongroup                1.1.3
gpg                           1.23.2
idna                          3.6
imagesize                     1.4.1
importlib_metadata            7.0.1
iniconfig                     2.0.0
installer                     0.7.0
Jinja2                        3.1.3
libdnf                        0.73.0
MarkupSafe                    2.1.3
packaging                     24.0
pluggy                        1.4.0
Pygments                      2.17.2
pyproject_hooks               1.0.0
pytest                        8.1.1
python-dateutil               2.9.0.post0
requests                      2.31.0
setuptools                    69.1.1
snowballstemmer               2.2.0
Sphinx                        7.2.6
sphinxcontrib-applehelp       1.0.8
sphinxcontrib-devhelp         1.0.5
sphinxcontrib-htmlhelp        2.0.5
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          1.0.7
sphinxcontrib-serializinghtml 1.1.10
strictyaml                    1.7.3
tokenize_rt                   5.2.0
tomli                         2.0.1
urllib3                       1.26.18
wheel                         0.43.0
zipp                          3.17.0
```
</details>

<details>
<summary>List of installed modules in build env:</summary>

```console
```
</details>

Please let me know if you need more details or want me to perform some diagnostics.

<details>
<summary>Here is pytest output:</summary>

```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-hawkmoth-0.17.0-2.fc36.x86_64/usr/lib64/python3.9/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-hawkmoth-0.17.0-2.fc36.x86_64/usr/lib/python3.9/site-packages
+ /usr/bin/pytest -ra -m 'not network'
============================= test session starts ==============================
platform linux -- Python 3.9.18, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/tkloczko/rpmbuild/BUILD/hawkmoth-0.17.0
configfile: pyproject.toml
testpaths: test
collected 272 items

test/test_cautodoc.py ..............x........................xF......... [ 18%]
................................x........................xF............. [ 44%]
..............                                                           [ 50%]
test/test_cli.py .ssssss.......x....s....ssssss.........xF..........s.ss [ 70%]
sssss......s.                                                            [ 75%]
test/test_parser.py ..............x........................xF........... [ 94%]
................                                                         [100%]

=================================== FAILURES ===================================
____________________ test_directive_text[cpp/mixed-linkage] ____________________

testcase = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc12ed60>

    @pytest.mark.parametrize('testcase', _get_extension_testcases(testenv.testdir, 'text'),
                             ids=testenv.get_testid)
    def test_directive_text(testcase):
>       testcase.run_test()

test/test_cautodoc.py:92:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc12ed60>

    def run_test(self):
        if self.options.get('expected-failure'):
            pytest.xfail()

        output_docs, output_errors = self.get_output()
        expect_docs, expect_errors = self.get_expected()

        assert output_docs, 'empty output'
        assert expect_docs, 'empty expected'
>       assert output_docs == expect_docs
E       AssertionError: assert 'int foo(void...+ function.\n' == 'int foo(void...+ function.\n'
E
E         Skipping 77 identical leading characters in diff, use -v to show
E           2.
E         -
E         - int bar(void)
E         -
E         -    C function....
E
E         ...Full output truncated (4 lines hidden), use '-vv' to show

test/testenv.py:113: AssertionError
----------------------------- Captured stdout call -----------------------------
Running Sphinx v7.2.6
making output directory... done
building [mo]: targets for 0 po files that are out of date
writing output...
building [text]: targets for 1 source files that are out of date
updating environment: [new config] 1 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... done
writing output... [100%] index
build succeeded, 1 warning.

The text files are in ../../../../../tmp/tmp1ckl00_9/text.
Running Sphinx v7.2.6
making output directory... done
building [mo]: targets for 0 po files that are out of date
writing output...
building [text]: targets for 1 source files that are out of date
updating environment: [new config] 1 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... done
writing output... [100%] index
build succeeded.

The text files are in ../../../../../tmp/tmp3da54g4k/text.
____________________ test_directive_html[cpp/mixed-linkage] ____________________

testcase = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc16ecd0>

    @pytest.mark.full
    @pytest.mark.parametrize('testcase', _get_extension_testcases(testenv.testdir, 'html'),
                             ids=testenv.get_testid)
    def test_directive_html(testcase):
>       testcase.run_test()

test/test_cautodoc.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc16ecd0>

    def run_test(self):
        if self.options.get('expected-failure'):
            pytest.xfail()

        output_docs, output_errors = self.get_output()
        expect_docs, expect_errors = self.get_expected()

        assert output_docs, 'empty output'
        assert expect_docs, 'empty expected'
>       assert output_docs == expect_docs
E       assert '<!DOCTYPE ht...ody>\n</html>' == '<!DOCTYPE ht...ody>\n</html>'
E
E         Skipping 1770 identical leading characters in diff, use -v to show
E         - l class="c function">
E         - <dt class="sig sig-object c" id="c.bar">
E         - <span class="kt"><span class="pre">int</span></span><span class="w"> </span><span class="sig-name descname"><span class="n"><span class="pre">bar</span></span></span><span class="sig-paren">(</span><span class="kt"><span class="pre">void</span></span><span class="sig-paren">)</span><a class="headerlink" href="#c.bar" title="Link to this definition">¶</a><br /></dt>
E         - <dd><p>C function.</p>
E         - </dd></dl>...
E
E         ...Full output truncated (27 lines hidden), use '-vv' to show

test/testenv.py:113: AssertionError
----------------------------- Captured stdout call -----------------------------
Running Sphinx v7.2.6
making output directory... done
building [mo]: targets for 0 po files that are out of date
writing output...
building [html]: targets for 1 source files that are out of date
updating environment: [new config] 1 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... copying static files... done
copying extra files... done
done
writing output... [100%] index
generating indices... done
writing additional pages... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in ../../../../../tmp/tmpqrfgpr8r/html.
Running Sphinx v7.2.6
making output directory... done
building [mo]: targets for 0 po files that are out of date
writing output...
building [html]: targets for 1 source files that are out of date
updating environment: [new config] 1 added, 0 changed, 0 removed
reading sources... [100%] index
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... copying static files... done
copying extra files... done
done
writing output... [100%] index
generating indices... done
writing additional pages... done
dumping object inventory... done
build succeeded.

The HTML pages are in ../../../../../tmp/tmpuvrme3ku/html.
_________________________ test_cli[cpp/mixed-linkage] __________________________

testcase = <test.test_cli.CliTestcase object at 0x7f48bc01a610>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f48b1d5e9a0>
capsys = <_pytest.capture.CaptureFixture object at 0x7f48b1d5ee80>

    @pytest.mark.full
    @pytest.mark.parametrize('testcase', _get_cli_testcases(testenv.testdir),
                             ids=testenv.get_testid)
    def test_cli(testcase, monkeypatch, capsys):
        testcase.set_monkeypatch(monkeypatch)
        testcase.set_capsys(capsys)
>       testcase.run_test()

test/test_cli.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cli.CliTestcase object at 0x7f48bc01a610>

    def run_test(self):
        if self.options.get('expected-failure'):
            pytest.xfail()

        output_docs, output_errors = self.get_output()
        expect_docs, expect_errors = self.get_expected()

        assert output_docs, 'empty output'
        assert expect_docs, 'empty expected'
>       assert output_docs == expect_docs
E       AssertionError: assert '\n.. cpp:fun...function.\n\n' == '\n.. cpp:fun...function.\n\n'
E
E         Skipping 97 identical leading characters in diff, use -v to show
E            2.
E         -
E         -
E         - .. c:function:: int bar(void)
E         - ...
E
E         ...Full output truncated (7 lines hidden), use '-vv' to show

test/testenv.py:113: AssertionError
________________________ test_parser[cpp/mixed-linkage] ________________________

testcase = <test.test_parser.ParserTestcase object at 0x7f48bc14b4c0>

    @pytest.mark.parametrize('testcase', _get_parser_testcases(testenv.testdir),
                             ids=testenv.get_testid)
    def test_parser(testcase):
>       testcase.run_test()

test/test_parser.py:124:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_parser.ParserTestcase object at 0x7f48bc14b4c0>

    def run_test(self):
        if self.options.get('expected-failure'):
            pytest.xfail()

        output_docs, output_errors = self.get_output()
        expect_docs, expect_errors = self.get_expected()

        assert output_docs, 'empty output'
        assert expect_docs, 'empty expected'
>       assert output_docs == expect_docs
E       AssertionError: assert '\n.. cpp:fun...function.\n\n' == '\n.. cpp:fun...function.\n\n'
E
E         Skipping 97 identical leading characters in diff, use -v to show
E            2.
E         -
E         -
E         - .. c:function:: int bar(void)
E         - ...
E
E         ...Full output truncated (7 lines hidden), use '-vv' to show

test/testenv.py:113: AssertionError
================================== XFAILURES ===================================
_________________ test_directive_text[c/meta-expected-failure] _________________

testcase = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc116fd0>

    @pytest.mark.parametrize('testcase', _get_extension_testcases(testenv.testdir, 'text'),
                             ids=testenv.get_testid)
    def test_directive_text(testcase):
>       testcase.run_test()

test/test_cautodoc.py:92:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc116fd0>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
________________ test_directive_text[cpp/meta-expected-failure] ________________

testcase = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc134520>

    @pytest.mark.parametrize('testcase', _get_extension_testcases(testenv.testdir, 'text'),
                             ids=testenv.get_testid)
    def test_directive_text(testcase):
>       testcase.run_test()

test/test_cautodoc.py:92:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc134520>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
_________________ test_directive_html[c/meta-expected-failure] _________________

testcase = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc109130>

    @pytest.mark.full
    @pytest.mark.parametrize('testcase', _get_extension_testcases(testenv.testdir, 'html'),
                             ids=testenv.get_testid)
    def test_directive_html(testcase):
>       testcase.run_test()

test/test_cautodoc.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc109130>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
________________ test_directive_html[cpp/meta-expected-failure] ________________

testcase = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc122a60>

    @pytest.mark.full
    @pytest.mark.parametrize('testcase', _get_extension_testcases(testenv.testdir, 'html'),
                             ids=testenv.get_testid)
    def test_directive_html(testcase):
>       testcase.run_test()

test/test_cautodoc.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cautodoc.ExtensionTestcase object at 0x7f48bc122a60>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
______________________ test_cli[c/meta-expected-failure] _______________________

testcase = <test.test_cli.CliTestcase object at 0x7f48bc097340>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f48b000c910>
capsys = <_pytest.capture.CaptureFixture object at 0x7f48b000c730>

    @pytest.mark.full
    @pytest.mark.parametrize('testcase', _get_cli_testcases(testenv.testdir),
                             ids=testenv.get_testid)
    def test_cli(testcase, monkeypatch, capsys):
        testcase.set_monkeypatch(monkeypatch)
        testcase.set_capsys(capsys)
>       testcase.run_test()

test/test_cli.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cli.CliTestcase object at 0x7f48bc097340>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
_____________________ test_cli[cpp/meta-expected-failure] ______________________

testcase = <test.test_cli.CliTestcase object at 0x7f48bc0d3ca0>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f48abf13ee0>
capsys = <_pytest.capture.CaptureFixture object at 0x7f48abf136a0>

    @pytest.mark.full
    @pytest.mark.parametrize('testcase', _get_cli_testcases(testenv.testdir),
                             ids=testenv.get_testid)
    def test_cli(testcase, monkeypatch, capsys):
        testcase.set_monkeypatch(monkeypatch)
        testcase.set_capsys(capsys)
>       testcase.run_test()

test/test_cli.py:99:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_cli.CliTestcase object at 0x7f48bc0d3ca0>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
_____________________ test_parser[c/meta-expected-failure] _____________________

testcase = <test.test_parser.ParserTestcase object at 0x7f48bbfeb220>

    @pytest.mark.parametrize('testcase', _get_parser_testcases(testenv.testdir),
                             ids=testenv.get_testid)
    def test_parser(testcase):
>       testcase.run_test()

test/test_parser.py:124:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_parser.ParserTestcase object at 0x7f48bbfeb220>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
____________________ test_parser[cpp/meta-expected-failure] ____________________

testcase = <test.test_parser.ParserTestcase object at 0x7f48bc14b820>

    @pytest.mark.parametrize('testcase', _get_parser_testcases(testenv.testdir),
                             ids=testenv.get_testid)
    def test_parser(testcase):
>       testcase.run_test()

test/test_parser.py:124:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <test.test_parser.ParserTestcase object at 0x7f48bc14b820>

    def run_test(self):
        if self.options.get('expected-failure'):
>           pytest.xfail()
E           _pytest.outcomes.XFailed

test/testenv.py:106: XFailed
=========================== short test summary info ============================
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc097940> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc0c7b80> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc1768e0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc148940> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc1487f0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc0aaf10> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc0c7790> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc1790a0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc179fa0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc176310> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc016c10> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc016b20> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc025fa0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc143430> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc0295e0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc17f070> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc17f520> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc03b0d0> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc0c7a30> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc029100> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc16ee80> directive test
SKIPPED [1] test/test_cli.py:59: <test.testenv.Directive object at 0x7f48bc016850> directive test
XFAIL test/test_cautodoc.py::test_directive_text[c/meta-expected-failure] - reason:
XFAIL test/test_cautodoc.py::test_directive_text[cpp/meta-expected-failure] - reason:
XFAIL test/test_cautodoc.py::test_directive_html[c/meta-expected-failure] - reason:
XFAIL test/test_cautodoc.py::test_directive_html[cpp/meta-expected-failure] - reason:
XFAIL test/test_cli.py::test_cli[c/meta-expected-failure] - reason:
XFAIL test/test_cli.py::test_cli[cpp/meta-expected-failure] - reason:
XFAIL test/test_parser.py::test_parser[c/meta-expected-failure] - reason:
XFAIL test/test_parser.py::test_parser[cpp/meta-expected-failure] - reason:
FAILED test/test_cautodoc.py::test_directive_text[cpp/mixed-linkage] - Assert...
FAILED test/test_cautodoc.py::test_directive_html[cpp/mixed-linkage] - assert...
FAILED test/test_cli.py::test_cli[cpp/mixed-linkage] - AssertionError: assert...
FAILED test/test_parser.py::test_parser[cpp/mixed-linkage] - AssertionError: ...
============ 4 failed, 238 passed, 22 skipped, 8 xfailed in 45.80s =============
```
</details>

I don't anything obvious in that output which can point on possible cause 🤔

--- a/src/hawkmoth/doccursor.py
+++ b/src/hawkmoth/doccursor.py
@@ -1,5 +1,5 @@
 # Copyright (c) 2016-2023 Jani Nikula <jani@nikula.org>
-# Copyright (c) 2018-2023 Bruno Santos <brunomanuelsantos@tecnico.ulisboa.pt>
+# Copyright (c) 2018-2024 Bruno Santos <brunomanuelsantos@tecnico.ulisboa.pt>
 # Licensed under the terms of BSD 2-Clause, see LICENSE for details.
 
 from clang.cindex import (
@@ -146,10 +146,11 @@ class DocCursor:
         """Get children cursors."""
         domain = self.domain
 
-        # Identify `extern "C"` blocks and change domain accordingly. For some
-        # reason, the Python bindings don't return the cursor kind LINKAGE_SPEC
-        # as one would expect, so we need to do it the hard way.
-        if domain == 'cpp' and self.kind == CursorKind.UNEXPOSED_DECL:
+        # Identify `extern "C"` blocks and change domain accordingly.
+        # Prior to Clang 18, the Python bindings don't return the cursor kind
+        # LINKAGE_SPEC as one would expect, so we need to do it the hard way.
+        if domain == 'cpp' and self.kind in [CursorKind.LINKAGE_SPEC,
+                                             CursorKind.UNEXPOSED_DECL]:
             tokens = self.get_tokens()
             ntoken = next(tokens, None)
             if ntoken and ntoken.spelling == 'extern':

--- a/src/hawkmoth/parser.py
+++ b/src/hawkmoth/parser.py
@@ -1,5 +1,5 @@
 # Copyright (c) 2016-2017 Jani Nikula <jani@nikula.org>
-# Copyright (c) 2018-2023 Bruno Santos <brunomanuelsantos@tecnico.ulisboa.pt>
+# Copyright (c) 2018-2024 Bruno Santos <brunomanuelsantos@tecnico.ulisboa.pt>
 # Licensed under the terms of BSD 2-Clause, see LICENSE for details.
 """
 Documentation comment extractor
@@ -174,7 +174,8 @@ def _comment_extract(tu):
 
         # Cursors that are 1) never documented themselves, and 2) not allowed
         # between the comment and the actual cursor being documented.
-        if token_cursor.kind in [CursorKind.UNEXPOSED_DECL]:
+        if token_cursor.kind in [CursorKind.LINKAGE_SPEC,
+                                 CursorKind.UNEXPOSED_DECL]:
             if is_doc(current_comment):
                 top_level_comments.append(current_comment)
             current_comment = None
@@ -299,9 +300,9 @@ def _parse_undocumented_block(errors, cursor, nest):
 
     # Identify `extern "C"` and `extern "C++"` blocks and recursively parse
     # their contents.
-    # For some reason, the Python bindings don't return the cursor kind
+    # Prior to Clang 18, the Python bindings don't return the cursor kind
     # LINKAGE_SPEC as one would expect, so we need to do it the hard way.
-    if cursor.kind == CursorKind.UNEXPOSED_DECL:
+    if cursor.kind in [CursorKind.LINKAGE_SPEC, CursorKind.UNEXPOSED_DECL]:
         tokens = cursor.get_tokens()
         ntoken = next(tokens, None)
         if ntoken and ntoken.spelling == 'extern':

